### PR TITLE
Never drop the `terminationGracePeriodSeconds` of a probe

### LINKED ISSUE #119377: Does not respect the terminationGracePeriodSeconds of a probe
### What happened?

Kubernetes does not respect the `terminationGracePeriodSeconds` of a probe.

/kind regression
/kind bug
/sig node

/priority important-soon

### What did you expect to happen?

The `terminationGracePeriodSeconds` of a container's probe should override the pod's `terminationGracePeriodSeconds`.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a pod with long pod-level terminationGracePeriodSeconds and short probe-level terminationGracePeriodSeconds.

```sh
$ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - name: regular-1
    image: registry.k8s.io/e2e-test-images/busybox:1.29-4
    startupProbe:
      exec:
        command:
        - /bin/false
      failureThreshold: 1
      terminationGracePeriodSeconds: 1
    command:
    - sh
    - -c
    - |
      sleep 300s
  terminationGracePeriodSeconds: 300
EOF
```

2. Check for the container not respecting the probe-level terminationGracePeriodSeconds
Every 2.0s: kubectl describe pods test | tail                                                                                                                                  gjkim-desktop: Tue Jul 18 00:19:23 2023

                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age    From               Message
  ----     ------     ----   ----               -------
  Normal   Scheduled  3m23s  default-scheduler  Successfully assigned default/test to kind-control-plane
  Normal   Pulled     3m22s  kubelet            Container image "registry.k8s.io/e2e-test-images/busybox:1.29-4" already present on machine
  Normal   Created    3m22s  kubelet            Created container regular-1
  Normal   Started    3m22s  kubelet            Started container regular-1
  Warning  Unhealthy  3m12s  kubelet            Startup probe failed:
  Normal   Killing    3m12s  kubelet            Container regular-1 failed startup probe, will be restarted

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?

/kind regression
/kind bug

#### What this PR does / why we need it:

This PR fixes the apiserver to never drop the `terminationGracePeriodSeconds` field of a probe

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/119377

#### Does this PR introduce a user-facing change?

```release-note
none
```
wrote none, as the bug is not released yet.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
> Passing "/bin/sh" arguments to agnhost container has caused failure by itself.

does this mean it never worked?

there was a great effort to move everything to agnhost , why do we switch to busybox?

**Comment by gjkim42** [comment]
> > Passing "/bin/sh" arguments to agnhost container has caused failure by itself.
> 
> does this mean it never worked?
> 
> there was a great effort to move everything to agnhost , why do we switch to busybox?

```
$ k run test --image=registry.k8s.io/e2e-test-images/agnhost:2.45 -- /bin/sh -c "echo hello"
pod/test created
$ k logs test
Error: unknown command "/bin/sh" for "app"
```

yes, we should not pass arguments to agnhost container.

**Comment by gjkim42** [comment]
```
$ k run test --image=registry.k8s.io/e2e-test-images/agnhost:2.45 --command -- /bin/sh -c "echo hello"
pod/test created
$ k logs test
hello
```

found that this works. we should pass the command as `command` not `args`

**Comment by gjkim42** [comment]
There may be a regression in recent changes.
https://github.com/kubernetes/kubernetes/issues/119377

File: pkg/api/pod/util.go [hunk 1]
```diff
@@ -516,20 +516,6 @@ func dropDisabledFields(
 		}
 	}
 
-	if !probeGracePeriodInUse(oldPodSpec) {
```
Comment (liggitt) [inline pkg/api/pod/util.go:519]: also drop the unused probeGracePeriodInUse method

File: test/e2e/common/node/container_probe.go [hunk 2]
```diff
@@ -460,13 +460,10 @@ var _ = SIGDescribe("Probing container", func() {
 		Description: A pod with a long terminationGracePeriod is created with a shorter livenessProbe-level terminationGracePeriodSeconds. We confirm the shorter termination period is used.
 	*/
 	ginkgo.It("should override timeoutGracePeriodSeconds when LivenessProbe field is set [NodeConformance]", func(ctx context.Context) {
-		pod := e2epod.NewAgnhostPod(f.Namespace.Name, "liveness-override-"+string(uuid.NewUUID()), nil, nil, nil, "/bin/sh", "-c", "sleep 1000")
```
[the remark this answers is not in the document]
Comment (gjkim42) [reply test/e2e/common/node/container_probe.go:463]: Actually, this proves that the timeoutGracePeriodSeconds field is not dropped and works as expected. Without fixing this e2e test, we cannot make sure that the regression is fixed.

[the remark this answers is not in the document]
Comment (liggitt) [reply test/e2e/common/node/container_probe.go:463]: a unit test would prove it and be simpler to reason about... this e2e change modifies things that are hard to understand

[the remark this answers is not in the document]
Comment (gjkim42) [reply test/e2e/common/node/container_probe.go:463]: I found this regression after fixing this e2e test.

[the remark this answers is not in the document]
Comment (gjkim42) [reply test/e2e/common/node/container_probe.go]: removed the e2e fix. let me know if you think we need a unit test for not dropping this field.
(I guess we usually test for dropping some fields.)

**Pull request merged**